### PR TITLE
Moving demotion/promotion passes to input conversion.

### LINKED ISSUE #8745: Fix issues with dialect conversion on linalg/arith/math ops for i64/f64 support.
The i64->i32 and f64->f32 demotion passes can cause verification failures in ops that are strict in verifiers but lacking folders. One such example is an `arith.extsi` i32 -> i64 that gets converted to an i32 -> i32:
```
D:\Dev\iree/iree/test/e2e/models/unidirectional_lstm.mlir:62:9: error: 'arith.extsi' op operand type 'i32' and result type 'i32' are cast incompatible
```

In order to support the type promotion/demotion the dialects we are working with need to have sufficient folders for these operations such that we can safely convert the types without tripping the verifier.

Currently only MHLO is properly supported for promotion/demotion as its ops lack verifiers and we can perform conversion successfully. This does not help TOSA/non-TOSA torch/direct-to-linalg/etc though and requires additional MHLO-specific input options instead of more generic compiler options. MHLO must currently default to demotion as otherwise the generic pass in flow cannot work.

To reproduce the above failure disable the MHLO demotion and just use the lower-level flow demotion:
```
iree-opt \
  -iree-mhlo-demote-i64-to-i32=false \
  -iree-flow-demote-i64-to-i32=true \
  -iree-import-public \
  -iree-mhlo-input-transformation-pipeline \
  -iree-abi-transformation-pipeline \
  -iree-flow-transformation-pipeline \
  iree/test/e2e/models/unidirectional_lstm.mlir
```

Part of this issue is scoping out what possible folders are needed upstream, whether we have other cases that come from tosa/etc that we need to support, etc. Ideally when fixed we completely remove the `-iree-mhlo-demote-*` flags and rely on the `-iree-flow-demote-*` flags instead.

### PR DESCRIPTION
**Pull request by benvanik**
They need to happen before ABI materialization. Really they should happen in frontend dialects where the semantics can be enumerated and verified. This at least removes the duplication we had between the pass flags in StableHLO and the global optimization flags that could conflict.

The f64->f32 demotion is still enabled by default because codegen has issues (CPU pulls in libm functions like exp, SPIR-V for all targets has issues with f64 memrefs, etc). At least now there's a single flag and it happens in a place that somewhat works and users can start to disable for testing.

StableHLO previously defaulted i64->i32 but all other input types didn't. The default is now i64->i32 disabled by default everywhere. People using StableHLO and wanting i64->i32 demotion will need to opt in to that behavior with `--iree-input-demote-i64-to-i32=true`.

Closes #8745.

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/Common/Passes.cpp [hunk 1]
```diff
@@ -7,35 +7,62 @@
...
-void buildCommonInputConversionPassPipeline(OpPassManager &passManager) {
+void buildCommonInputConversionPassPipeline(
+    OpPassManager &passManager, const TransformOptions &transformOptions) {
   passManager.addPass(createIREEImportPublicPass());
   passManager.addPass(createImportMLProgramPass());
   passManager.addPass(createSanitizeModuleNamesPass());
+
+  // TODO: this pass should either live in InputConversion or be run in flow -
+  // it's a mistake that it's here.
   passManager.addPass(IREE::Flow::createConvertMeshToFlowPass());
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/InputConversion/Common/Passes.cpp:31]: Outdated TODO? (This is in InputConversion)

**Review by ScottTodd — APPROVED** [review]

File: compiler/src/iree/compiler/InputConversion/Common/Passes.cpp (hunk 1 above)
Comment (benvanik) [reply compiler/src/iree/compiler/InputConversion/Common/Passes.cpp:31]: New TODO - the pass lives in flow but is added in input conversion - it should either live in InputConversion or be run in flow. I have no clue why it was done like this.

Comment (ScottTodd) [reply compiler/src/iree/compiler/InputConversion/Common/Passes.cpp:31]: Ah, gotcha.

**Comment by benvanik** [comment]
Oh, fun, stablehlo defaulted i64->i32 but all other input types didn't. The default is now i64->i32 disabled by default everywhere. People using stablehlo and wanting i64->i32 demotion will need to opt in to that behavior.

**Pull request merged**